src/transformers/models/timesfm/modeling_timesfm.py

```diff
@@ -0,0 +1,562 @@
+            hidden_size=self.config.model_dim,
+            intermediate_size=self.config.model_dim,
+            num_heads=self.config.num_heads,
+            num_kv_heads=self.config.num_heads,
+            head_dim=self.config.head_dim,
+            num_layers=self.config.num_layers,
+            rms_norm_eps=self.config.rms_norm_eps,
+        )
+        if self.config.use_positional_embedding:
+            self.position_emb = PositionalEmbedding(
+                embedding_dims=self.config.model_dim,
+            )
+
+    def _forward_transform(
+        self, inputs: torch.Tensor, patched_pads: torch.Tensor
+    ) -> tuple[torch.Tensor, tuple[torch.Tensor, torch.Tensor]]:
+        """Input is of shape [B, N, P]."""
+        mu, sigma = masked_mean_std(inputs, patched_pads)
+        sigma = torch.where(
+            sigma < self.config.tolerance,
+            torch.tensor(1.0, dtype=sigma.dtype, device=sigma.device),
+            sigma,
+        )
+
+        # Normalize each patch
+        outputs = (inputs - mu[:, None, None]) / sigma[:, None, None]
+        outputs = torch.where(
+            torch.abs(inputs - self.config.pad_val) < self.config.tolerance,
+            torch.tensor(
+                self.config.pad_val, dtype=outputs.dtype, device=outputs.device
+            ),
+            outputs,
+        )
+        return outputs, (mu, sigma)
+
+    def _reverse_transform(
+        self, outputs: torch.Tensor, stats: tuple[torch.Tensor, torch.Tensor]
+    ) -> torch.Tensor:
+        """Output is of shape [B, N, P, Q]."""
+        mu, sigma = stats
+        return outputs * sigma[:, None, None, None] + mu[:, None, None, None]
+
+    def _preprocess_input(
+        self,
+        input_ts: torch.Tensor,
+        input_padding: torch.Tensor,
+    ) -> tuple[
+        torch.Tensor,
+        torch.Tensor,
+        tuple[torch.Tensor, torch.Tensor] | None,
+        torch.Tensor,
+    ]:
+        """Preprocess input for stacked transformer."""
+
+        # Reshape into patches (using view for efficiency)
+        bsize = input_ts.shape[0]
+        patched_inputs = input_ts.view(bsize, -1, self.config.patch_len)
+        patched_pads = input_padding.view(bsize, -1, self.config.patch_len)
+
+        patched_inputs = torch.where(
+            torch.abs(patched_pads - 1.0) < self.config.tolerance,
+            torch.tensor(0.0, dtype=patched_inputs.dtype, device=patched_inputs.device),
+            patched_inputs,
+        )
+        patched_pads = torch.where(
+            torch.abs(patched_inputs - self.config.pad_val) < self.config.tolerance,
+            torch.tensor(1.0, dtype=patched_pads.dtype, device=patched_pads.device),
+            patched_pads,
+        )
+        patched_inputs, stats = self._forward_transform(patched_inputs, patched_pads)
+
+        # B x N x D
+        patched_inputs = patched_inputs * (1.0 - patched_pads)
+        print(">>> PatchedDecoder patched_inputs", patched_inputs.shape)
```

Comment:
nit: not sure if these print statements are necessary.

Reply:
The PR is still in progress. We will make sure to eliminate all debug info when ready for review.